src/VisualStudio/Core/Def/Implementation/VirtualMemoryNotificationListener.cs

```diff
@@ -104,5 +116,46 @@ public int OnBroadcastMessage(uint msg, IntPtr wParam, IntPtr lParam)
 
             return VSConstants.S_OK;
         }
+
+        private bool ShouldTurnOffFullSolutionAnalysis(long availableMemory)
+        {
+            // conditions
+            // 1. if available memory is less than the threshold and 
+            // 2. if full solution analysis memory monitor is on (user can set it off using registery, when he does, we will never show info bar) and
+            // 3. if our full solution analysis option is on (not user full solution analysis option, but our internal one) and
+            // 4. if infobar is never shown to users for this solution
+            return availableMemory < MemoryThreshold &&
+                  _workspace.Options.GetOption(InternalFeatureOnOffOptions.FullSolutionAnalysisMemoryMonitor) &&
+                  _workspace.Options.GetOption(RuntimeOptions.FullSolutionAnalysis) &&
+                  !_workspace.Options.GetOption(RuntimeOptions.FullSolutionAnalysisInfoBarShown);
+        }
+
+        private bool IsUserOptionOn()
+        {
+            // check languages currently on solution. since we only show info bar once, we don't need to track solution changes.
+            var languages = _workspace.CurrentSolution.Projects.Select(p => p.Language).Distinct();
+            foreach (var language in languages)
+            {
+                if (ServiceFeatureOnOffOptions.IsClosedFileDiagnosticsEnabled(_workspace, language))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        {
+            if (e.Kind != WorkspaceChangeKind.SolutionAdded)
+            {
+                return;
+            }
+
+            // first make sure full solution analysis is on. (not user options but our internal options. even if our option is on, if user option is off
```

Comment:
We discussed this offline and @heejaechang mentioned that design team decided to keep the turning off scope to be the current solution for the current VS instance, not for entire lifetime of this VS instance. In that case, this is fine, but we should make this clear in the README link.